decoder.go

```diff
@@ -960,6 +957,24 @@ func (d *Decoder) Skip(tag int, wt WireType) ([]byte, error) {
 	return d.p[bof:d.offset], nil
 }
 
+// validateFieldLen is a helper to DRY up checking the result of DecodeVarint() when used to read the
+// length of a length-delimited field.
+func (d *Decoder) validateFieldLen(l uint64, n int, err error) error {
+	// length-delimited fields cannot contain more than 2GB
+	const maxFieldLen = math.MaxInt32
+
+	switch {
+	case err != nil:
+		return fmt.Errorf("invalid data at byte %d: %w", d.offset, err)
```

Comment:
It seems odd to me that an error check for an error that occurs outside this function is checked here. Is this done for a performance reason or just to prevent another if err != nil...

Reply:
This is DRY'ing up checking the 3 failure cases for `DecodeVarint()`when the value is the length of a length-delimited field.  I'll admit it's a bit odd, but it felt less odd than checking 1 of the 3 failure cases at each call site.

Reply:
I ended up removing the `validateFieldLen()` helper.  It did feel just a bit too unidiomatic for the benefit it was providing